modules/cudaimgproc/src/cuda/hough_lines.cu

```diff
@@ -174,33 +171,30 @@ namespace cv { namespace cuda { namespace device
             }
         }
 
-        int linesGetResult_gpu(PtrStepSzi accum, float2* out, int* votes, int maxSize, float rho, float theta, int threshold, bool doSort)
+        int linesGetResult_gpu(PtrStepSzi accum, float2* out, int* votes, int maxSize, float rho, float theta, int threshold, bool doSort, int* counterPtr, cudaStream_t stream)
         {
-            void* counterPtr;
-            cudaSafeCall( cudaGetSymbolAddress(&counterPtr, g_counter) );
-
-            cudaSafeCall( cudaMemset(counterPtr, 0, sizeof(int)) );
+            cudaSafeCall( cudaMemsetAsync(counterPtr, 0, sizeof(int), stream) );
 
             const dim3 block(32, 8);
             const dim3 grid(divUp(accum.cols - 2, block.x), divUp(accum.rows - 2, block.y));
 
             cudaSafeCall( cudaFuncSetCacheConfig(linesGetResult, cudaFuncCachePreferL1) );
 
-            linesGetResult<<<grid, block>>>(accum, out, votes, maxSize, rho, theta, threshold, accum.cols - 2);
+            linesGetResult<<<grid, block, 0, stream>>>(accum, out, votes, maxSize, rho, theta, threshold, accum.cols - 2, counterPtr);
             cudaSafeCall( cudaGetLastError() );
 
-            cudaSafeCall( cudaDeviceSynchronize() );
-
             int totalCount;
-            cudaSafeCall( cudaMemcpy(&totalCount, counterPtr, sizeof(int), cudaMemcpyDeviceToHost) );
+            cudaSafeCall( cudaMemcpyAsync(&totalCount, counterPtr, sizeof(int), cudaMemcpyDeviceToHost, stream) );
+
+            cudaSafeCall( cudaStreamSynchronize(stream) );
 
             totalCount = ::min(totalCount, maxSize);
 
             if (doSort && totalCount > 0)
             {
                 thrust::device_ptr<float2> outPtr(out);
                 thrust::device_ptr<int> votesPtr(votes);
-                thrust::sort_by_key(votesPtr, votesPtr + totalCount, outPtr, thrust::greater<int>());
+                thrust::sort_by_key(thrust::cuda::par.on(stream), votesPtr, votesPtr + totalCount, outPtr, thrust::greater<int>());
```

Comment:
Could you please take a look if there is an easy way to guard this code for CUDA 8.x?

Reply:
I do not understand your question here.
Is there a problem specific to CUDA 8 ? I have checked the API and it seems fine.
As Cuda 8 is not supported anymore on Ubuntu 18.04 or Visual Studio 19 (my dev tools),
I took some time today to setup an Ubuntu 16.04/CUDA 8 and it builds properly with OpenCV 4.5.0 (master did not on some unrelated problem).

Reply:
Please take a look on build issues of "Custom" CUDA builder [here](http://pullrequest.opencv.org/#/summary/contrib).

There are several ways:
- try to fix the build using `#if` guards with CUDA version checks (requested above)
- try to fix the build through disabling this algorithm for CUDA 8.0
- open a discussion about deprecation of old CUDA versions in next OpenCV releases (will take some time)

Reply:
It should be fixed on my last commit.
I managed to reproduce the build problem, the instruction on the NVIDIA website to install cuda 8 were wrong, I should have checked.
Turns out that the thrust::cuda::par did exist in CUDA 8, but was not included by default.